controllers/restore_controller.go

```diff
@@ -19,133 +19,370 @@ package controllers
-		return ctrl.Result{}, errors.Wrap(r.Client.Status().Update(ctx, restore), "could not update status")
+	switch {
+	case len(veleroRestoreList.Items) == 0:
+		veleroRestore, err := r.initVeleroRestore(ctx, restore)
+		if err != nil {
+			restoreLogger.Error(err, "unable to initialize velero restore for restore %s/%s", req.Namespace, req.Name)
+			return ctrl.Result{}, err
+		}
+		if err = r.Create(ctx, veleroRestore, &client.CreateOptions{}); err != nil {
+			restoreLogger.Error(err, "unable to create velero restore for restore %s/%s", req.Namespace, req.Name)
+			return ctrl.Result{}, err
+		}
+		r.Recorder.Event(restore, v1.EventTypeNormal, "Velero Restore created:", fmt.Sprintf("%s/%s", veleroRestore.Namespace, veleroRestore.Name))
+		restore.Status.VeleroRestoreName = veleroRestore.Name
+		apimeta.SetStatusCondition(&restore.Status.Conditions,
+			metav1.Condition{
+				Type:    v1beta1.RestoreStarted,
+				Status:  metav1.ConditionTrue,
+				Reason:  v1beta1.RestoreReasonStarted,
+				Message: fmt.Sprintf("Velero Restore %s Started", veleroRestore.Name),
+			})
+	case len(veleroRestoreList.Items) == 1:
+		veleroRestore := veleroRestoreList.Items[0].DeepCopy()
+		switch {
+		case isRestoreFinished(restore):
+			return ctrl.Result{}, nil // won't do more
+		case isVeleroRestoreFinished(veleroRestore):
+			r.Recorder.Event(restore, v1.EventTypeNormal, "Velero Restore finished:", fmt.Sprintf("%s/%s", veleroRestore.Namespace, veleroRestore.Name))
+			allAttached, err := r.reattachManagedClusters(ctx)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			switch {
+			case allAttached:
+				apimeta.SetStatusCondition(&restore.Status.Conditions,
+					metav1.Condition{
+						Type:    v1beta1.RestoreComplete,
+						Status:  metav1.ConditionTrue,
+						Reason:  v1beta1.RestoreReasonRunning,
+						Message: fmt.Sprintf("Restore Complete %s", restore.Name),
+					})
+				return ctrl.Result{}, r.Client.Status().Update(ctx, restore)
+			case !allAttached:
+				apimeta.SetStatusCondition(&restore.Status.Conditions,
+					metav1.Condition{
+						Type:    v1beta1.RestoreReattaching,
+						Status:  metav1.ConditionTrue,
+						Reason:  v1beta1.RestoreReasonRunning,
+						Message: fmt.Sprintf("Restore %s is re-attaching clusters", restore.Name),
+					})
+				return ctrl.Result{RequeueAfter: time.Second * 10}, r.Client.Status().Update(ctx, restore) // to wait for managed cluster re-attaching phase...
+			}
+		default:
+			apimeta.SetStatusCondition(&restore.Status.Conditions,
+				metav1.Condition{
+					Type:    v1beta1.RestoreStarted,
+					Status:  metav1.ConditionTrue,
+					Reason:  v1beta1.RestoreReasonRunning,
+					Message: fmt.Sprintf("Velero Restore %s still running", veleroRestore.Name),
+				})
+		}
+
+	default:
+		// TODO: handles multiple velero restores:
+		// check if one velero is still running... update status and wait
+		// if all finished handleManagedClusters
+		//for i := range veleroRestoreList.Items {}
```

Comment:
maybe remove this line

Reply:
Sure, it's there for me. But you're right